File: app/assets/javascripts/helpers/schoolYear.js
import moment from 'moment';

// year: An integer year.
// returns: A moment object representing the first day of the school year that year starts.
export function firstDayOfSchool(year) {
  return toMoment([year, 8, 15]);
}

// year: An integer year.
// returns: A moment object representing roughly the last day of that school year (which will
// be in the following calendar year).
export function lastDayOfSchool(year) {
  return toMoment([year + 1, 6, 30]);
}

// Returns moment representing the first day of school for a given moment.
export function firstDayOfSchoolForMoment(someMoment) {
  const year = toSchoolYear(someMoment.toDate());
  return firstDayOfSchool(year);
}

// date: A JS date object or Moment object.
// returns: Integer representing what the calendar year was in the fall of date's school year.
export function toSchoolYear(date) {
  const momentObject = moment.utc(date);

  const year = momentObject.year();
  const startOfSchoolYear = firstDayOfSchool(year);
  const isEventDuringFall = momentObject.diff(startOfSchoolYear, 'days') > 0;
  return (isEventDuringFall) ? year : year - 1;
}

function toMoment(triple) {
  return moment.utc([triple[0], triple[1], triple[2]].join('-'), 'YYYY-M-D');
}

File: app/lib/school_year.rb
# Define when school years start and end.  See also schoolYear.js.
class SchoolYear
  # date: A JS date object or Moment object.
  # returns: Integer representing what the calendar year was in the fall of date's school year.
  def self.first_day_of_school_for_year(year)
    DateTime.new(year, 8, 15)
  end

  # year: An integer year.
  # returns: A moment object representing roughly the last day of that school year (which will
  # be in the following calendar year).
  def self.last_day_of_school_for_year(year)
    DateTime.new(year + 1, 6, 30)
  end

  # Returns DateTime representing the first day of school for the given date_time
  def self.first_day_of_school_for_time(date_time)
    year = to_school_year(date_time)
    first_day_of_school_for_year(year)
  end

  # returns: Integer representing what the calendar year was in the fall of date's school year
  def self.to_school_year(date_time)
    year = date_time.year
    start_of_school_year = self.first_day_of_school_for_year(year)
    is_event_during_fall = (date_time - start_of_school_year) > 0
    if is_event_during_fall
      year
    else
      year - 1
    end
  end
end
